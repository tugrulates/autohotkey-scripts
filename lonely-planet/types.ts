/** A Lonely Planet destination type. */
export type DestionationType =
  | "Continent"
  | "Country"
  | "Region"
  | "City"
  | "Neighborhood";

/** A Lonely Planet attraction type. */
export type AttractionType = "Attractions";

/** A Lonely Planet story type. */
export type StoryType = "Feature" | "News";

/** A Lonely Planet document. */
export interface Document {
  /** Document ID. */
  id: string;
  /** Document URL. */
  slug: string;
  /** Document title. */
  title: string;
  /** Summary for topic of the document. */
  excerpt: string;
  /** Document image. */
  featuredImage: Image;
}

/** A Lonely Planet destination. */
export interface Destination extends Document {
  /** Destination type. */
  type: DestionationType;
  /** Global breadcrumb to the location. */
  breadcrumb: Breadcrumb[];
}

/** A Lonely Planet attraction. */
export interface Attraction extends Document {
  /** Attraction type. */
  type: AttractionType;
  /** Global breadcrumb to the location. */
  breadcrumb: Breadcrumb[];
}

/** A Lonely Planet story. */
export interface Story extends Document {
  /** Story type. */
  type: StoryType;
  /** Story publication date in ISO string. */
  date: string;
  /** Story read time in minutes. */
  readTime: number;
}

/** Global path component of a Lonely Planet document. */
export interface Breadcrumb {
  /** Node URL. */
  slug: string;
  /** Node title. */
  title: string;
  /** Node type. */
  type: DestionationType | AttractionType;
}

/** A Lonely Planet image. */
export interface Image {
  /** Alt text for the image. */
  alt: string;
  /** Caption for the image. */
  caption: string;
  /** Image credit. */
  credit: string;
  /** Image width in pixels. */
  width: number;
  /** Image height in pixels. */
  height: number;
  /** Image URL. */
  title: string;
  /** Image URL. */
  url: string;
}

/** A results page for a query on the Lonely Planet. */
export interface SearchResults {
  /** Number of results in this page. */
  found: number;
  /** Number of results in all pages. */
  out_of: number;
  /** The current page number. */
  page: number;
  /** Results and scoring info. */
  hits: [
    {
      document: Destination | Attraction | Story;
      text_match: number;
      text_match_info: {
        best_field_score: string;
        best_field_weight: number;
        fields_matched: number;
        num_tokens_dropped: number;
        score: string;
        tokens_matched: number;
        typo_prefix_score: number;
      };
    },
  ];
  /** The query parameters. */
  request_params: {
    collection_name: string;
    first_q: string;
    per_page: number;
    q: string;
  };
  /** ? */
  search_cutoff: boolean;
  /** Time it took to retrieve the results. */
  search_time_ms: number;
}

/** A Typesense server error. */
export interface Error {
  /** HTTP response code. */
  code: number;
  /** Error message. */
  error: string;
}
