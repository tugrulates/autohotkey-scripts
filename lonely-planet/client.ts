import { toPascalCase } from "jsr:@std/text";
import * as uuid from "jsr:@std/uuid";
import { JsonClient } from "../common/request.ts";
import { Attraction, Destination, SearchResults, Story } from "./types.ts";

/**
 * Client for interacting with the Lonely Planet API.
 *
 * Requires credentials to the Typesense API, which can be obtained from the Lonely Planet frontend.
 */
export class LonelyPlanetClient {
  private client: JsonClient;

  /**
   * Creates an instance of the Lonely Planet client.
   *
   * @param endpoint The endpoint for the Lonely Planet API.
   * @param token The Typesense API key.
   */
  constructor(endpoint: string, private token: string) {
    this.client = new JsonClient(endpoint);
  }

  /**
   * Searches Lonely Planet for destinations with the given keywords.
   *
   * @param keywords The keywords to search for.
   * @returns Generator for the search results.
   */
  async *searchDestionations(keywords: string[]): AsyncGenerator<Destination> {
    for await (
      const document of this.search<Destination>("places", keywords)
    ) {
      yield document as Destination;
    }
  }

  /**
   * Searches Lonely Planet for attractions with the given keywords.
   *
   * @param keywords The keywords to search for.
   * @returns Generator for the search results.
   */
  async *searchAttractions(keywords: string[]): AsyncGenerator<Attraction> {
    for await (const document of this.search<Attraction>("pois", keywords)) {
      yield document as Attraction;
    }
  }

  /**
   * Searches Lonely Planet for stories with the given keywords.
   *
   * @param keywords The keywords to search for.
   * @returns Generator for the search results.
   */
  async *searchStories(keywords: string[]): AsyncGenerator<Story> {
    for await (const document of this.search<Story>("articles", keywords)) {
      yield document as Story;
    }
  }

  /**
   * Searches Lonely Planet for the given keywords.
   *
   * @template T The type of document to search for.
   * @param collection The collection to search.
   * @param keywords The keywords to search for.
   * @returns Generator for the search results.
   */
  private async *search<T extends Destination | Attraction | Story>(
    collection: "places" | "pois" | "articles",
    keywords: string[],
  ): AsyncGenerator<T> {
    let page = 1;
    while (true) {
      const body = {
        searches: [{
          preset: `global_search_${collection}`,
          collection,
          q: keywords.join(" "),
          per_page: 100,
          page,
        }],
      };
      const results = (await this.client.post<
        { results: (SearchResults | Error)[] }
      >(
        `/multi_search?x-typesense-api-key=${this.token}`,
        body,
      )).results[0];
      if ("error" in results) {
        throw new Error(`{data.results[0].error}`);
      } else if ("hits" in results) {
        for (const hit of results.hits) {
          const document = hit.document;
          // ignore documents from old ID formats, as these are duplicates
          if (uuid.validate(document.id)) {
            // Cleanup types.
            let type = toPascalCase(document.type);
            if (type === "" && "breadcrumb" in document) {
              const parentType = document.breadcrumb.at(-1)?.type;
              if (parentType === "Continent") type = "Country";
              else if (parentType === "Country") type = "Region";
            }
            yield {
              ...document,
              type,
            } as T;
          }
        }
        if (!results.hits.length) break;
        page++;
      } else {
        break;
      }
    }
  }
}
