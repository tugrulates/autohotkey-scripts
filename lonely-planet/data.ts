/** Lonely Planet document type emojis. */
export const EMOJIS = {
  Continent: "🌍",
  Country: "🏳️",
  Region: "🏞️",
  City: "🏙️",
  Neighborhood: "🏘️",
  Attractions: "🎡",
  Feature: "📰",
  News: "📢",
};
