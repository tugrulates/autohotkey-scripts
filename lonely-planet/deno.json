{
  "name": "@tugrulates/lonely-planet",
  "version": "0.1.2",
  "exports": "./mod.ts",
  "compile": {
    "main": "mod.ts",
    "permissions": {
      "net": true
    }
  }
}
