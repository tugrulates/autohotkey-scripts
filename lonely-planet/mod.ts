/**
 * Interact with Lonely Planet, the travel guide website.
 *
 * @module
 */

import { main } from "./main.ts";

export * from "./client.ts";
export * from "./data.ts";
export * from "./types.ts";

if (import.meta.main) await main(Deno.args);
