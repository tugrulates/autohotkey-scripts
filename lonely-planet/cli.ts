import { Command } from "jsr:@cliffy/command";
import { Table } from "jsr:@cliffy/table";
import { Config } from "../common/cli.ts";
import { LonelyPlanetClient } from "./client.ts";
import { EMOJIS } from "./data.ts";
import type { Attraction, Destination } from "./types.ts";

const endpoint = new Config("endpoint");
const token = new Config("token", { secret: true });

/** Lonely Planet client built from common CLI options. */
export async function getClient(): Promise<LonelyPlanetClient> {
  return new LonelyPlanetClient(await endpoint.get(), await token.get());
}

/** Breadcrumb text from a location. */
function breadcrumb(document: Destination | Attraction) {
  return `[ ${
    document.breadcrumb.map((breadcrumb) => breadcrumb.title).join(" > ")
  } ]`;
}

/**
 * Command line interface for Lonely Planet.
 *
 * @ignore missing-explicit-type
 */
export const command = new Command()
  .name("lonely-planet")
  .description("Explores data from Lonely Planet.")
  .example("lonely-planet big sur", "Search destinations for 'big sur'.")
  .example("lonely-planet --attractions amsterdam", "Search attractions.")
  .example("lonely-planet --stories amsterdam", "Search stories.")
  .example("lonely-planet --destinations --attractions --stories", "All.")
  .example("lonely-planet --json | jq", "Stream destinations as json.")
  .option(
    "--endpoint <endpoint:string>",
    "Typesense endpoint for Lonely Planet.",
    await endpoint.option(),
  )
  .option(
    "--token <token:string>",
    "Typesense token for Lonely Planet.",
    await token.option(),
  )
  .arguments("[...keywords:string]")
  .option("--destinations", "Include destinations in the results.")
  .option("--attractions", "Include attractions in the results.")
  .option("--stories", "Include stories in the results.")
  .option("--json", "Output the search results as concatenated JSON.")
  .action(async ({ destinations, attractions, stories, json }, ...keywords) => {
    if (!destinations && !attractions && !stories) destinations = true;
    const client = await getClient();
    const rows: string[][] = [];
    if (destinations) {
      for await (const doc of client.searchDestionations(keywords)) {
        if (json) console.log(JSON.stringify(doc));
        else rows.push([EMOJIS[doc.type], doc.title, breadcrumb(doc)]);
      }
    }
    if (attractions) {
      for await (const doc of client.searchAttractions(keywords)) {
        if (json) console.log(JSON.stringify(doc));
        else rows.push([EMOJIS[doc.type], doc.title, breadcrumb(doc)]);
      }
    }
    if (stories) {
      for await (const doc of client.searchStories(keywords)) {
        if (json) console.log(JSON.stringify(doc));
        else rows.push([EMOJIS[doc.type], doc.title]);
      }
    }
    Table.from(rows).render();
  });

if (import.meta.main) {
  await command.parse();
}
