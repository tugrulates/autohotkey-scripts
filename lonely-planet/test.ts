import { assertSnapshot } from "@std/testing/snapshot";
import { LonelyPlanetClient } from "@tugrulates/lonely-planet";
import { mockFetch } from "@tugrulates/testing";

Deno.test("searchDestinations", async (t) => {
  using _fetch = mockFetch(t);
  const client = new LonelyPlanetClient();
  const results = await Array.fromAsync(
    client.searchDestinations(["big", "sur"]),
  );
  await assertSnapshot(t, results);
});

Deno.test("searchAttractions", async (t) => {
  using _fetch = mockFetch(t);
  const client = new LonelyPlanetClient();
  const results = await Array.fromAsync(
    client.searchAttractions(["big", "sur"]),
  );
  await assertSnapshot(t, results);
});

Deno.test("searchStories", async (t) => {
  using _fetch = mockFetch(t);
  const client = new LonelyPlanetClient();
  const results = await Array.fromAsync(
    client.searchStories(["big", "sur"]),
  );
  await assertSnapshot(t, results);
});
