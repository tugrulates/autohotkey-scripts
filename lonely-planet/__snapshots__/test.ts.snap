export const snapshot = {};

snapshot[`searchDestinations 1`] = `
[
  {
    breadcrumb: [
      {
        slug: "north-america",
        title: "North America",
        type: "Continent",
      },
      {
        slug: "usa",
        title: "USA",
        type: "Country",
      },
      {
        slug: "usa/california",
        title: "California",
        type: "Region",
      },
    ],
    excerpt: "Big Sur is more a state of mind than a place to pinpoint on a map, and when the sun goes down, the moon and the stars are the area's natural streetlights. (That's if summer’s fog hasn’t extinguished them.) Raw beauty and an intense maritime energy characterize this land shoehorned between the Santa Lucia Range and the Pacific Ocean, and a first glimpse of the craggy, unspoiled coastline is a special moment.",
    featuredImage: {
      alt: "Bixby Bridge on Highway One in Big Sur, CA",
      caption: "Bixby Bridge on Highway One in Big Sur, CA",
      credit: null,
      height: 2679,
      name: "28206231.jpg",
      title: "28206231.jpg",
      url: "https://lp-cms-production.imgix.net/2019-06/28206231.jpg",
      width: 4034,
    },
    id: "150856",
    internalRanking: 1.77918,
    slug: "usa/california/big-sur",
    title: "Big Sur",
    type: "Region",
  },
]
`;

snapshot[`searchAttractions 1`] = `
[
  {
    breadcrumb: [
      {
        slug: "north-america",
        title: "North America",
        type: "Continent",
      },
      {
        slug: "usa",
        title: "USA",
        type: "Country",
      },
      {
        slug: "usa/california",
        title: "California",
        type: "Region",
      },
      {
        slug: "usa/california/big-sur/attractions",
        title: "Attractions",
        type: "Attractions",
      },
    ],
    excerpt: "Big Sur's largest state park was named after the area's first European settlers who arrived in 1869. It has hiking trails looping through stately redwood…",
    featuredImage: {},
    id: "3794286",
    internalRanking: 0.55,
    slug: "usa/california/big-sur/attractions/pfeiffer-big-sur-state-park/a/poi-sig/470074/361791",
    title: "Pfeiffer Big Sur State Park",
    type: "Attractions",
  },
  {
    breadcrumb: [
      {
        slug: "north-america",
        title: "North America",
        type: "Continent",
      },
      {
        slug: "usa",
        title: "USA",
        type: "Country",
      },
      {
        slug: "usa/california",
        title: "California",
        type: "Region",
      },
      {
        slug: "usa/california/big-sur/attractions",
        title: "Attractions",
        type: "Attractions",
      },
    ],
    excerpt: "From the parking lot at Andrew Molera State Park, walk 0.125 miles south to this educational nature center to learn all about the endangered California…",
    featuredImage: {},
    id: "3794991",
    internalRanking: 0.434,
    slug: "usa/california/big-sur/attractions/big-sur-discovery-center/a/poi-sig/1253133/361791",
    title: "Big Sur Discovery Center",
    type: "Attractions",
  },
]
`;

snapshot[`searchStories 1`] = `
[
  {
    date: "2024-12-19T15:00:00.000Z",
    excerpt: "Watch the land and ocean play out their endless drama at these Big Sur beaches",
    featuredImage: {
      alt: null,
      caption: "Enjoy all the drama the sea can offer at these Big Sur beaches © Alexander Spatari / Getty Images",
      credit: "© Alexander Spatari / Getty Images",
      height: 3840,
      name: "Cliffs and ocean at Big Sur, California, USA",
      title: "Cliffs and ocean at Big Sur, California, USA",
      url: "https://lp-cms-production.imgix.net/2021-09/Beach%20in%20Big%20Sur.jpg",
      width: 5760,
    },
    id: "10359661",
    readTime: 11,
    slug: "articles/best-beaches-in-big-sur",
    title: "11 bucket-list Big Sur beaches",
    type: "Feature",
  },
  {
    date: "2021-09-16T16:25:23.000Z",
    excerpt: "This 90-mile stretch of California coastline offers tidepooling, whale watching, camping, music events and more. And then there are the beaches.",
    featuredImage: {
      alt: "Coastline on Highway 1 near Big Sur.",
      caption: "Hwy 1 through Big Sur, California, is one of the country’s most scenic routes © Kan Khampanya /500px",
      credit: "©Kan Khampanya /500px",
      height: 3090,
      name: "500px Photo ID: 7949546 - I took this shot on Highway 1 on the way from Big Sur back to my hotel. The sun was setting and wild flowers beside the road make us pull off our car and took some pictures here..",
      title: "500px Photo ID: 7949546 - I took this shot on Highway 1 on the way from Big Sur back to my hotel. The sun was setting and wild flowers beside the road make us pull off our car and took some pictures here..",
      url: "https://lp-cms-production.imgix.net/2020-11/500pxRF_7949546.jpg",
      width: 5000,
    },
    id: "10359921",
    readTime: "8",
    slug: "articles/best-time-to-visit-big-sur",
    title: "The best time to visit Big Sur: why it's always in season",
    type: "Feature",
  },
  {
    date: "2021-09-16T16:14:12.000Z",
    excerpt: "With its distinct dining options, and fun nature and historical experiences, this California jewel shines.",
    featuredImage: {
      alt: "Bixby Bridge, California",
      caption: "You can't reach Big Sur, California, without crossing the Bixby Bridge, which has been in movies, car commercials and TV shows © Ozéias Sant'ana / 500px",
      credit: "©Ozéias Sant'ana/500px",
      height: 3648,
      name: "Bixby Bridge from the eastern side",
      title: "Bixby Bridge from the eastern side",
      url: "https://lp-cms-production.imgix.net/2020-11/500pxRF_107278777.jpg",
      width: 5472,
    },
    id: "10359981",
    readTime: "5",
    slug: "articles/top-things-to-do-in-big-sur",
    title: "In Big Sur, you can do it all - or nothing",
    type: "Feature",
  },
  {
    date: "2021-09-08T17:00:00.000Z",
    excerpt: "Big Sur is more than just a famous drive. This scenic stretch of coast has hikes for all levels and views to die for. Here are the best Big Sur hikes.",
    featuredImage: {
      alt: "Bixby Bridge, California",
      caption: "Bixby Bridge, California",
      credit: "©Ozéias Sant'ana/500px",
      height: 3648,
      name: "Bixby Bridge from the eastern side",
      title: "Bixby Bridge from the eastern side",
      url: "https://lp-cms-production.imgix.net/2020-11/500pxRF_107278777.jpg",
      width: 5472,
    },
    id: "10359836",
    readTime: 7,
    slug: "articles/best-hikes-in-big-sur",
    title: "7 Big Sur hikes to get you out of the car and onto the trails",
    type: "Feature",
  },
  {
    date: "2018-11-15T13:18:53.000Z",
    excerpt: "Along California’s scenic Highway 1, nestled high on wild mountain cliffs overlooking the crashing waves of the Pacific Ocean lie the picturesque…",
    featuredImage: {
      alt: "a finished landscape stands on an easel in with wildflowers and ocean in the background",
      caption: "Art is a way of life along the Central Coast of California © Chris Cheadle / Getty Images",
      credit: "",
      height: 1053,
      name: "Features - Carmel.Art_.Hero_-d96412487b16",
      title: "",
      url: "https://lp-cms-production.imgix.net/features/2018/11/Carmel.Art_.Hero_-d96412487b16.jpg",
      width: 1600,
    },
    id: "10038461",
    readTime: 5,
    slug: "articles/chasing-art-carmel-big-sur",
    title: "Chasing art in Carmel-by-the-Sea and Big Sur",
    type: "Feature",
  },
  {
    date: "2018-10-22T21:37:36.000Z",
    excerpt: "An epic road trip is one of America’s favorite pastimes, and few drives are more iconic than Highway 1 through Big Sur. Tight turns hug the rugged…",
    featuredImage: {
      alt: "Blooming purple flowers sway in the breeze backed by a steep cliff coastline",
      caption: "The best season for wildflowers is spring, but any time is right for a drive down Highway 1. Allard Schager / Getty Images",
      credit: "©© Allard Schager/Getty Images",
      height: 2832,
      name: "media:canto_dam:2fe9af1a-5346-437a-b7f3-573938f46ae3",
      title: "media:canto_dam:2fe9af1a-5346-437a-b7f3-573938f46ae3",
      url: "https://lp-cms-production.imgix.net/2024-10/GettyRF623466212.jpg",
      width: 4256,
    },
    id: "10038466",
    readTime: 5,
    slug: "articles/scenic-big-sur-top-stops-highway-1-dream-drive",
    title: "Scenic Big Sur: top sights on your Highway 1 road trip",
    type: "Feature",
  },
  {
    date: "2018-01-25T17:40:25.000Z",
    excerpt: "For decades travelers have journeyed to Big Sur seeking communion with the wild. Artists and writers, adventure-seekers and soul-searchers have looked to…",
    featuredImage: {
      alt: "Highway one snakes around a cliff boarding the pacific ocean",
      caption: "Highway one snakes around a cliff boarding the pacific ocean",
      credit: "",
      height: 3532,
      name: "Features - First-image-80122d92ec4b",
      title: "",
      url: "https://lp-cms-production.imgix.net/features/2018/01/First-image-80122d92ec4b.jpg",
      width: 5298,
    },
    id: "10038456",
    readTime: 5,
    slug: "articles/escape-to-big-sur-two-days-to-reconnect-body-mind-and-spirit",
    title: "Weekend escape to Big Sur",
    type: "Feature",
  },
  {
    date: "2012-07-04T09:45:22.000Z",
    excerpt: "You're exhausted, disoriented and in need of exercise. Your throat is dry. You smell like a gymnasium changing room and you oscillate between feeling…",
    featuredImage: {},
    id: "10023846",
    readTime: "3",
    slug: "articles/planes-trains-and-automobiles-how-to-survive-a-big-trip",
    title: "Planes, trains and automobiles: how to survive a big trip",
    type: "Feature",
  },
]
`;
